src/tariff_region.h

```diff
@@ -0,0 +1,131 @@
+#ifndef CYCAMORE_SRC_TARIFF_REGION_H_
+#define CYCAMORE_SRC_TARIFF_REGION_H_
+
+#include "cyclus.h"
+#include <string>
+#include <algorithm>
+#include <limits>
+#include <set>
+#include <utility>
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+namespace cycamore {
+
+// Simple structure to represent a tariff rule
+struct TariffRule {
+  std::string region_name;
+  std::string commodity;
+  double adjustment;
+  
+  TariffRule(const std::string& region, const std::string& comm, double adj)
+      : region_name(region), commodity(comm), adjustment(adj) {}
+};
+
+class TariffRegion : public cyclus::Region {
+ public:
+  TariffRegion(cyclus::Context* ctx);
+  virtual ~TariffRegion();
+
+  virtual void EnterNotify();
+  virtual void Tock();
+
+  // Required DRE Functions
+  virtual void AdjustMatlPrefs(cyclus::PrefMap<cyclus::Material>::type& prefs);
+  virtual void AdjustProductPrefs(cyclus::PrefMap<cyclus::Product>::type& prefs);
+
+ private:
+  // Build the set of region agents for faster lookups
+  void BuildRegionSet();
+  
+  // Find the best matching region for a supplier
+  cyclus::Region* FindMatchingRegion(cyclus::Facility* supplier);
+  
+  // Find the appropriate tariff for a given region and commodity
+  double FindTariffForCommodity(cyclus::Region* region, const std::string& commodity);
+  
+  // Build tariff rules from the input configuration
+  void BuildTariffRules();
+  
+  // Validate the tariff configuration
+  void ValidateConfiguration();
+  
+  // Record tariff configuration to database
+  void RecordTariffConfiguration();
+
+  #pragma cyclus
+
+  // Template function to reduce code duplication between AdjustMatlPrefs and 
+  // AdjustProductPrefs
+  template<typename T>
+  void AdjustPrefsImpl(typename cyclus::PrefMap<T>::type& prefs) {
+    for (auto& req_pair : prefs) {
+      std::string commodity = req_pair.first->commodity();
+      
+      for (auto& bid_pair : req_pair.second) {
+        cyclus::Bid<T>* bid = bid_pair.first;
+        cyclus::Facility* supplier = dynamic_cast<cyclus::Facility*>(bid->bidder()->manager());
+        
+        // Find if any of the supplier's parent regions match our tariff list
+        cyclus::Region* matching_region = FindMatchingRegion(supplier);
+        if (matching_region) {
+          double adjustment = FindTariffForCommodity(matching_region, commodity);
+          
+          double cost_multiplier = 1.0 + adjustment;
+          double pref_multiplier = 1.0 / cost_multiplier;
+          double inf = std::numeric_limits<double>::infinity(); 
+
+          bid_pair.second *= cost_multiplier > 0.0 ? pref_multiplier : inf; 
+        }
+      }
+    }
+  }
```

Comment:
These often live outside the the class declaration, at the end of this file, appropriately scoped, i.e. `void TariffRegion::AdjustPrefsImpl(...`